### PR TITLE
Support cancel job with savepoint

### PR DESCRIPTION
**Pull request by functicons**
Add a property `cancelRequested` in jobSpec and allow it to be updated.

### CODE REVIEW

File: api/v1alpha1/flinkcluster_validate.go [hunk 1]
```diff
@@ -55,14 +55,45 @@ func (v *Validator) ValidateCreate(cluster *FlinkCluster) error {
...
+func (v *Validator) checkCancelRequested(
+	old *FlinkCluster, new *FlinkCluster) (bool, error) {
+	if old.Spec.Job == nil || new.Spec.Job == nil {
+		return false, nil
+	}
+	var restartJob = (old.Spec.Job.CancelRequested != nil && *old.Spec.Job.CancelRequested) &&
+		(new.Spec.Job.CancelRequested == nil || !*new.Spec.Job.CancelRequested)
+	if restartJob {
+		return false, fmt.Errorf(
+			"updating cancelRequested from true to false is not allowed")
```
Comment (hongyegong) [inline api/v1alpha1/flinkcluster_validate.go]: Why cant this be just restart is not allowed after cancle?

File: controllers/flinkcluster_reconciler.go [hunk 2]
```diff
@@ -428,23 +424,50 @@ func (reconciler *ClusterReconciler) getFlinkJobID() string {
...
 	}
+	if !reconciler.isJobFinished() {
+		return reconciler.stopJob(jobID)
+	}
+	log.Info("Job has finished, no need to cancel", "jobID", jobID)
+	return nil
+}
+
+// Stops a job through Flink REST API.
+func (reconciler *ClusterReconciler) stopJob(jobID string) error {
```
Comment (hongyegong) [inline controllers/flinkcluster_reconciler.go]: Worth creating a wrapper method here? seems like it's only called once in this method and whatever pre-logic here can just go there.

File: controllers/flinkcluster_updater.go [hunk 3]
```diff
@@ -371,13 +374,19 @@ func (updater *ClusterStatusUpdater) deriveClusterStatus(
...
 			status.Components.Job.State = v1alpha1.JobState.Succeeded
 			jobFinished = true
 			jobSucceeded = true
 		} else {
 			status.Components.Job = nil
 		}
+	} else if recordedJobStatus != nil && len(recordedJobStatus.Name) > 0 {
+		status.Components.Job = recordedJobStatus.DeepCopy()
+		status.Components.Job.State = v1alpha1.JobState.Cancelled
+		jobFinished = true
```
Comment (hongyegong) [inline controllers/flinkcluster_updater.go]: Jobfinished doesn't seem to be homogeneous with jobcancled.

File: api/v1alpha1/flinkcluster_validate.go (hunk 1 above)
Comment (functicons) [reply api/v1alpha1/flinkcluster_validate.go]: We'll support restart after finishing #84.

**Comment by functicons** [comment]
Fix #79

**Pull request merged**